### PR TITLE
Implement git-based hashing with inputs

### PR DESCRIPTION
**Pull request by gsoltis**
### Description

 - replaces internal file hashing functions with top-level, public interface file hashing functions
 - moves remaining Go implementations into the Go-specific implementation file
 - introduces a `Hasher` struct to memoize the decision on git vs manual hashing, as well as some git details (we could consider directly executing the binary vs finding it every time)
 - removes `write_escaped_bytes` from `RelativeUnixPath` because we only needed it for writing to `git hash-object`. We now use `libgit2` directly and don't need to write to another process

### Testing Instructions

 - Switches Rust tests over to use the public functions
 - Adds test for hash-files

### CODE REVIEW

File: crates/turborepo-paths/src/relative_unix_path_buf.rs [hunk 1]
```diff
@@ -42,7 +42,11 @@ impl RelativeUnixPathBuf {
             .filter(|(_, byte)| **byte == b'\"' || **byte == b'\n')
         {
             writer.write_all(&self.0[i..to_escaped_index])?;
-            writer.write_all(&[b'\\', *byte])?;
+            if *byte == b'\n' {
+                writer.write_all(&[b'\\', b'n'])?;
```
Comment (gsoltis) [inline crates/turborepo-paths/src/relative_unix_path_buf.rs:46]: This is a bug fix, but now that we're using `libgit2`, I don't think we need this code at all

**Comment by gsoltis** [comment]
Previous test failures were due to misconfigured CI

File: crates/turborepo-paths/src/relative_unix_path_buf.rs (hunk 1 above)
Comment (gsoltis) [reply crates/turborepo-paths/src/relative_unix_path_buf.rs]: Removed this code.

**Review by NicholasLYang — APPROVED** [review]

**Comment by arlyon** [comment]
Could you please add the motivation for removing `write_escaped_bytes` to the PR description for the sake of posterity?

**Review by arlyon — APPROVED** [review]

File: crates/turborepo-scm/src/package_deps.rs [hunk 2]
```diff
@@ -3,84 +3,180 @@ use std::{collections::HashMap, process::Command};
...
-    let mut hashes = GitHashes::new();
-    hash_objects(&git_root, &full_pkg_path, to_hash, &mut hashes)?;
-    Ok(hashes)
+#[derive(Debug)]
+pub enum Hasher {
+    Git(Git),
+    Manual,
+}
+
+fn find_git(path_in_repo: &AbsoluteSystemPath) -> Option<Git> {
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-scm/src/package_deps.rs:26]: Yup, good suggestion

File: crates/turborepo-scm/src/package_deps.rs [hunk 3]
```diff
@@ -3,84 +3,180 @@ use std::{collections::HashMap, process::Command};
...
+        turbo_root: &AbsoluteSystemPath,
+        files: impl Iterator<Item = AnchoredSystemPathBuf>,
+    ) -> Result<GitHashes, Error> {
+        match self {
+            Hasher::Manual => crate::manual::hash_files(turbo_root, files, false),
+            Hasher::Git(git) => git.hash_files(turbo_root, files),
+        }
+    }
+
+    pub fn hash_existing_of(
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-scm/src/package_deps.rs:67]: It's probably not the best name, and copied from a Go name that is also probably not great. I'll add a comment

**Pull request merged**